### ISSUE TITLE
[mac-v2] (conditional) Layer-#5 sd-scripts patches — only if #22 perf target missed

### ISSUE THREAD

**Issue by hiyukoim** [hiyukoim/CoppyLora_webUI#23]
## Goal

**This issue is blocked on #22.** Open it (move out of `blocked` label) only if #22's smoke test shows steady-state step time still **>20 s/iter** on M4 Pro after #16-#21 are merged and system hygiene followed. If #22 passes, close this as `not_planned`.

If unblocked: apply targeted patches inside `sd-scripts/` to address platform-inherent overhead. Each patch must pass the **Quality Firewall** (defined below) before merging.

## Background

v1 → v2 (#16-#21) optimizations are all platform-layer: env vars, memory fraction, attention kernel selection, profile choice. They reduce memory pressure and replace the default attention path. They do **not** address:

- The fact that all training runs at fp32 (no fp16/bf16 mixed precision on MPS).
- The fact that AdamW32 carries 8× the optimizer state of AdamW8bit (which is unavailable on MPS).
- The fact that no 8-bit kernel exists for SDXL on Apple Silicon.

If those make the difference between 12 s/iter and 20 s/iter, layer #5 is unnecessary. If they don't, we have to dig into sd-scripts.

## Quality firewall (gate for any patch in this issue)

A patch lands **only if** all of the following hold:

1. It does not change `loss = ...`, optimizer step, gradient computation, weight init, noise sampling, or LR schedule.
2. A 50-step diff: same seed + image + caption produces a `.safetensors` whose **per-tensor max abs diff vs. v1 baseline is ≤ 1e-4**. >1e-4 means we changed math; revert.
3. A perceptual check: the v1 LoRA and the patched LoRA, loaded in Mac ComfyUI with the same prompt + seed, produce visually indistinguishable outputs.

The PR for any patch must include the diff numbers and the side-by-side ComfyUI screenshot.

### Patch A: `torch.mps.empty_cache()` between epochs

In sd-scripts' main training loop, call `torch.mps.empty_cache()` at the end of each epoch (or every N steps if epoch is too coarse). Memory hint, not math. Safe.

Files: likely `sd-scripts/library/train_util.py` or `sd-scripts/sdxl_train_network.py`'s training loop. Find the per-epoch hook and inject:

```python
if hasattr(torch, "mps") and torch.backends.mps.is_available():
    torch.mps.empty_cache()
```

### Patch B: Force SDPA math backend

The PyTorch SDPA kernel selection on MPS picks the EFFICIENT backend by default; on some workloads the MATH backend is faster. Wrap UNet attention calls in:

```python
from torch.nn.attention import SDPBackend, sdpa_kernel
with sdpa_kernel(SDPBackend.MATH):
    out = F.scaled_dot_product_attention(q, k, v, ...)
```

This is bit-identical math output (just a different kernel implementation), so the firewall is trivially passed if MPS-MATH is bit-identical to the default. Verify with the 50-step diff.

### Patch C: Pin `accelerate` to a known-good MPS version

If accelerate's MPS handling has regressed in a recent release, pin it to a known-good version in `install.sh`. Pure version pin, no code change. Check `pip show accelerate` after a fresh install.

### Patch D: `mps-bitsandbytes` integration (highest risk)

Try installing `mps-bitsandbytes` from PyPI and switching `optimizer_type = "AdamW8bit"`. **This will likely fail the firewall** because 8-bit optimizer state changes float precision. If the 50-step diff passes (unlikely), great — huge memory win. If not, revert and don't ship.

## What is explicitly off-limits

- `mixed_precision = "fp16"` or `"bf16"` (proven NaN on MPS).
- Replacing AdamW with a different optimizer in the **default** profile.
- Touching `library/strategy_*` or `library/sdxl_*` files (training math).

## Acceptance criteria

- [ ] #22 smoke test confirmed step time >20 s/iter and swap <2 GB (i.e. memory wasn't the bottleneck).
- [ ] At least one patch from A-C lands and passes the firewall.
- [ ] PR includes 50-step diff numbers and ComfyUI side-by-side.
- [ ] Step time after the patch is measurably lower than #22 baseline.
- [ ] `v1.2.0-mac` tag is cut after merge.

## Process

Each patch is its own atomic PR. We don't bundle. If patch A passes firewall and patch B doesn't, A ships and B is closed without merging.

## Out of scope

- Upstream PRs back to kohya-ss/sd-scripts (could be a follow-up — might benefit other Mac users).
- Patching diffusers itself (out of project scope).
